File: Resolution Algorithms Book/algorithms/binarySearch.js
export default function binarySearch(arr, target) {
    let start = 0;
    let end = arr.length - 1;
    while(start <= end){
        let middle = Math.floor((start + end) / 2)
        let guess = arr[middle]
        if(target === guess) return middle
        if(target > guess) { start  = middle + 1 }
        if(target < guess) {end = middle - 1}
    } 
    return null
}

//Exercise 1:
/*Suponha que você tenha uma lista com 128 nomes e esteja fazendo uma
pesquisa binária. Qual seria o número máximo de etapas que você levaria
para encontrar o nome desejado?*/
//Resposta: 7 etapas.

/*Suponha que você duplique o tamanho da lista. Qual seria o número
máximo de etapas agora?*/
//Resposta: 8 etapas porque log2(128) = 256, 2^8 = 256 e 2^7 = 128 que eh o mesmo que log2(256)=8.

File: Resolution Algorithms Book/algorithms/binary_search.cpp
#include <iostream>
#include <variant>
#include <cmath>

using namespace std;

int binary_search(int val[], int arr_size, int target){
	
	int start = 0;
	int end = arr_size;
	
	while(start <= end){
		int middle = (start + end) / 2;
		if(middle == target){return middle - 1;}
		if(target > middle){ start = middle + 1; }
		if(target < middle){ end = middle - 1; }
	}
	return -1;
}

int main() {
	int values[]{1, 2, 3, 4, 5, 6, 7, 8, 9};
	int arr_size = sizeof(values) / sizeof(values[0]);
	int target = 6;
	cout << binary_search(values, arr_size, target);
    return 0;
}




